contrib/groffer/roff2.1.man
.TH roff2@ROFF2MODE@ @MAN1EXT@ "@MDATE@" "groff @VERSION@"
.SH Name
roff2@ROFF2MODE@ \- transform roff code into @ROFF2MODE@ mode
.
.
.\" Save and disable compatibility mode (for, e.g., Solaris 10/11).
.do nr *groff_roff2@ROFF2MODE@_1_man_C \n[.cp]
.cp 0
.
.
.\" ====================================================================
.\" Legal Terms
.\" ====================================================================
.\"
.\" Copyright (C) 2006-2018 Free Software Foundation, Inc.
.\"
.\" This file is part of groffer, which is part of GNU groff, a free
.\" software project.
.\"
.\" You can redistribute it and/or modify it under the terms of the GNU
.\" General Public License version 2 (GPL2) as published by the Free
.\" Software Foundation.
.\"
.\" The license text is available in the internet at
.\" <http://www.gnu.org/licenses/gpl-2.0.html>.
.
.
.\" ====================================================================
.\" Characters
.\" ====================================================================
.
.\" Ellipsis ...
.ie t .ds EL \fS\N'188'\fP\"
.el .ds EL \&.\|.\|.\&\"
.\" called with \*(EL
.
.
.\" ====================================================================
.SH Synopsis
.\" ====================================================================
.
.SY roff2@ROFF2MODE@
.RI [ groffer-option
\*(EL]
.OP \-\-
.RI [ filespec
\*(EL]
.YS
.
.SY roff2@ROFF2MODE@
.B \-h
.SY roff2@ROFF2MODE@
.B \-\-help
.YS
.
.SY roff2@ROFF2MODE@
.B \-v
.SY roff2@ROFF2MODE@
.B \-\-version
.YS
.
.P
The options
.B \-v
and
.B \%\-\-version
print the version information of the program to standard output and exit.
.
The options
.B \-h
and
.B \-\-help
print a usage information of the program to standard output and stop
the program instantly.
.
.
.P
All other options are assumed to be
.B \%groffer
options.
.
They are internally passed to
.BR \%groffer .
They override the behavior of the program.
.
The options are optional, they can be omitted.
.
.
.P
The
.I \%filespec
arguments correspond to the
.I \%filespec
arguments of
.BR \%groffer .
So they are either the names of existing, readable files or
.B \-
for standard input, or the name of a man page or a
.BR \%groffer (@MAN1EXT@)
man page search pattern.
.
If no
.I \%filespec
is specified standard input is assumed automatically.
.
.
.\" ====================================================================
.SH Description
.\" ====================================================================
.
.B \%roff2@ROFF2MODE@
transforms
.I roff
code into
.ie '@ROFF2MODE@'x' \{\
.  I X
mode corresponding to the
.  I groff
devices
.  BR X *;
this mode is suitable for
.  BR \%gxditview (@MAN1EXT@).
.\}
.el \{\
.  I \%@ROFF2MODE@
mode.
.\}
.
Print the result to standard output.
.
.
.P
There are more of these programs for generating other formats of
.I \%roff
input.
.
.if !'@ROFF2MODE@'dvi' \{\
.  TP
.  BR \%roff2dvi (@MAN1EXT@)
is for
.  I dvi
mode.
.\}
.
.if !'@ROFF2MODE@'html' \{\
.  TP
.  BR \%roff2html (@MAN1EXT@)
generates
.  I HTML
output.
.\}
.
.if !'@ROFF2MODE@'pdf' \{\
.  TP
.  BR \%roff2pdf (@MAN1EXT@)
outputs
.  I pdf
mode.
.\}
.
.if !'@ROFF2MODE@'ps' \{\
.  TP
.  BR \%roff2ps (@MAN1EXT@)
prints
.  I \%PostScript
format to standard output.
.\}
.
.if !'@ROFF2MODE@'text' \{\
.  TP
.  BR \%roff2text (@MAN1EXT@)
generates text output in the
.  B groff
device
.  IR latin1 .
.\}
.
.if !'@ROFF2MODE@'x' \{\
.  TP
.  BR \%roff2x (@MAN1EXT@)
prints the output in the
.  B groff
device
.  B X
that is suitable for programs like
.  BR \%gxditview (@MAN1EXT@)
or
.  BR \%xditview (1).
.\}
.
.
.\" ====================================================================
.SH Authors
.\" ====================================================================
.B roff2@ROFF2MODE@
was written by
.MT groff\-bernd.warken\-72@\:web.de
Bernd Warken
.ME .
.
.
.\" ====================================================================
.SH "See Also"
.\" ====================================================================
.
.BR \%groff (@MAN1EXT@),
.BR \%groffer (@MAN1EXT@),
.if !'@ROFF2MODE@'dvi' \
.  BR \%roff2dvi (@MAN1EXT@),
.if !'@ROFF2MODE@'html' \
.  BR \%roff2html (@MAN1EXT@),
.if !'@ROFF2MODE@'pdf' \
.  BR \%roff2pdf (@MAN1EXT@),
.if !'@ROFF2MODE@'ps' \
.  BR \%roff2ps (@MAN1EXT@),
.if !'@ROFF2MODE@'text' \
.  BR \%roff2text (@MAN1EXT@),
.if !'@ROFF2MODE@'x' \
.  BR \%roff2x (@MAN1EXT@),
.BR \%gxditview (@MAN1EXT@).
.
.
.\" Restore compatibility mode (for, e.g., Solaris 10/11).
.cp \n[*groff_roff2@ROFF2MODE@_1_man_C]
.
.
.\" Local Variables:
.\" mode: nroff
.\" fill-column: 72
.\" End:
.\" vim: set filetype=groff textwidth=72:


contrib/groffer/roff2.pl
#! /usr/bin/env perl

# roff2* - transform roff files into other formats

# Source file position: <groff-source>/contrib/groffer/roff2.pl
# Installed position: <prefix>/bin/roff2*

# Copyright (C) 2006-2018 Free Software Foundation, Inc.
# Written by Bernd Warken <groff-bernd.warken-72@web.de>.

# This file is part of 'groffer', which is part of 'groff'.

# 'groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# 'groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/gpl-2.0.html>.

########################################################################

require v5.6;

use strict;
use warnings;
use File::Spec;

# temporary dir and files
use File::Temp qw/ tempfile /;

my $Dev_Null = File::Spec->devnull();
my $Mode;
my $Name;
{
  my ($v, $d);
  ($v, $d, $Name) = File::Spec->splitpath($0);
  die "wrong program name: $Name;"
    if $Name !~ /^roff2[a-z]/;
}
$Mode = $Name;
$Mode =~ s/^roff2//;

my $Groff_Version = `groff --version 2>$Dev_Null`;
die "$Name error: groff does not work;" if $?;
my $Groffer_Version = `groffer --version 2>$Dev_Null`;
my $Has_Groffer = ! $?;

if ($Has_Groffer) {
  foreach (@ARGV) {
    last if $_ eq '--';
    next if $_ eq '-';
    if ($_ eq '-v' || '--version' =~ m|^$_|) {
      print $Name, ' in ', $Groffer_Version;
      exit 0;
    }
    if ($_ eq '-h' || '--help' =~ m|^$_|) {
      &usage_with_groffer();
      exit 0;
    }
  }
  system('groffer', '--to-stdout', "--$Mode", @ARGV);
  exit $?;
} else {
  my @filespec;
  my $double_minus;
  foreach (@ARGV) {
    if ($double_minus) {
      push @filespec, $_;
      next;
    }
    if ($_ eq '--') {
      $double_minus = 1;
      next;
    }
    if ($_ eq '-') {
      push @filespec, '-';
      next;
    }
    if ($_ eq '-v' || '--version' =~ m|^$_|) {
      print $Name, ' in ', $Groff_Version;
      exit 0;
    }
    if ($_ eq '-h' || '--help' =~ m|^$_|) {
      &usage_without_groffer();
      exit 0;
    }
    if ($_ =~ /^-/) {
      &error_no_groffer();
      &error_no_options();
      next;
    }
    if (-f && -r) {
      push @filespec, $_;
    } else {
      &error_no_groffer();
      print STDERR "$_ is not an existing, readable file.\n";
    }
  }

  @filespec = ('-') unless @filespec;
  my $has_stdin;
  foreach (@filespec) {
    if ($_ eq '-') {
      $has_stdin =1;
      last;
    }
  }

  if ($has_stdin) {
    my $tempdir;
    foreach ($ENV{'GROFF_TMPDIR'}, $ENV{'TMPDIR'}, $ENV{'TMP'},
	     $ENV{'TEMP'}, $ENV{'TEMPDIR'},
	     File::Spec->catfile($ENV{'HOME'}, 'tmp')) {
      if ($_ && -d $_ && -w $_) {
	$tempdir = $_;
	last;
      }
    }
    my $template = $Name . '_XXXX';
    my ($fh, $stdin);
    if ($tempdir) {
      ($fh, $stdin) = tempfile($template, UNLINK => 1, DIR => $tempdir) ||
	die "$Name: could not create temporary file;";
    } else {
      ($fh, $stdin) = tempfile($template, UNLINK => 1) ||
	die "$Name: could not create temporary file;";
    }
    open $fh, ">$stdin";
    print $fh $_ foreach <STDIN>;

    foreach (@filespec) {
      $_ = $stdin if $_ eq '-';
    }
  }				# if $has_stdin

  my $grog;
  my $groff_options = "-T$Mode";
  $groff_options = '-TX75-12 -Z' if $Mode eq 'x';
  $groff_options = '-Tlatin1' if $Mode eq 'text';
  if ($Mode eq 'pdf') {
    my $ps2pdf;
    my @path = File::Spec->path();
    foreach (@path) {
      my $file = File::Spec->catfile($_, 'ps2pdf');
      if (-f $file && -x $file) {
	$ps2pdf = "$file -";
	last;
      }
    }
    unless ($ps2pdf) {
      foreach (@path) {
	my $file = File::Spec->catfile($_, 'gs');
	if (-f $file && -x $file) {
	  $ps2pdf = $file . ' -q -dNOPAUSE -dBATCH -sDEVICE=pdfwrite ' .
	    '-sOutputFile=- -c save pop -f -';
	  last;
	}
      }
    }
    $grog = `grog -Tps @filespec`;
    chomp $grog;
    system("$grog | $ps2pdf");
    exit $?;
  } else {
    $grog = `grog $groff_options @filespec`;
    chomp $grog;
    system($grog);
    exit $?;
  }
}

my $error_no_groffer;
sub error_no_groffer {
  return 1 if $error_no_groffer;
  $error_no_groffer = 1;
  print STDERR "$Name: groffer is not available.\n";
}

my $error_no_options;
sub error_no_options {
  return 1 if $error_no_options;
  $error_no_groffer = 1;
  print STDERR "$Name: groffer options are not allowed.\n";
}

sub usage_with_groffer {
  print <<EOF;
usage: $Name [option]... [--] [filespec]...

-h | --help     print usage information
-v | --version  print version information

All other options are arbitrary options of 'groffer'; the options
override the behavior of this program.

filespec's are the same as in 'groffer': either the names of existing,
readable files or '-' for standard input or a search pattern for man
pages.  No 'filespec' assumes standard input automatically.
EOF
}

sub usage_without_groffer {
  print <<EOF;
usage: $Name [option]... [--] [filespec]...

-h | --help     print usage information
-v | --version  print version information

No other options are allowed because 'groffer' is not available.

The only 'filespec's allowed are the names of existing, readable files
or '-' for standard input.  No 'filespec' assumes standard input
automatically.
EOF
}


1;
# Local Variables:
# mode: CPerl
# End:
